libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1065,6 +1065,87 @@ string YulUtilFunctions::clearStorageArrayFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::copyArrayToStorage(ArrayType const& _fromType, ArrayType const& _toType)
+{
+	solAssert(
+		*_fromType.copyForLocation(_toType.location(), _toType.isPointer()).get() == dynamic_cast<ReferenceType const&>(_toType),
+		""
+	);
+
+	string functionName = "copy_array_to_storage_from_" + _fromType.identifier() + "_to_" + _toType.identifier();
+	return m_functionCollector.createFunction(functionName, [&](){
+		Whiskers templ(R"(
```

Comment:
I think this template has too many case distinctions, we should try to simplify it...

For example, instead of conditionally having ``length`` for dynamic calldata arrays, we could use ``<value>`` and make it a list of ``_fromType.sizeOnStack()``-many variables, and then implement a ``arrayLengthFunction`` for dynamic calldata types as well (i.e. for *all* array types - which would be weird, since it'd be ``function arrayLengthFunction(value, len) -> length { length := len }`` in the dynamic calldata case, but it shouldn't harm, because it'll be inlined by the optimiser and in places like this it would make it easier to do things more generically).

Also I think it might be worth it to treat the ``isBytes`` case separately.

Since after the first suggestion the shared code for determining the length will reduce to one line in all cases, I think also splitting into "array of reference type" and "array of value type" cases could also work without much duplication.

Reply:
I simplified length part of template.
I tried to simplify other parts a bit without moving all cases to separate template, as it would have a lot of duplication than.
Does it look better now? @ekpyron

Reply:
I posted another comment about using the index access functions below, that will probably change the situation quite a bit...

I think if we want to do this efficiently, the first distinction will have to be based on whether the to-array will have multiple elements per slot or not. If it has multiple ones, we'll want to loop over slots and not elements, treating the last slot special. (e.g. as long as there's more than two ``uint128``s left, read two of them from calldata, pack them together and write to storage once, and only in the end check if a single one remains that has to be appended).

If it is one element per slot, on the other hand, it's much simpler.

In both cases for the source-side in memory/calldata we first calculate the starting offset before the loop and then just increment by ``calldataStride()`` or ``memoryStride()`` in the loop.

So none of the bounds checks hidden in the index access functions have to be performed inside the loop and for packed storage arrays, we can drastically reduce the storage reads and writes required, if multiple elements are written at once - which is worth the complexity, since storage access is really expensive...

Reply:
So doing all of this manually and most efficiently instead will probably be quite some work and tricky to get right - but I still think we should do it, since savings in storage operations are quite important and the optimizer has little chance here.

Reply:
I completely agree with your comment here. I'll leave this PR now in draft, and will do optimizations.
I'll use this code as a reference and incrementally add optimization and split into possibly more functions for different cases.

Reply:
I'll mark this as resolved now, but just because it's annoying to have the long comment in the sources ;-).